chalk-solve/src/clauses.rs

```diff
@@ -314,12 +314,15 @@ fn program_clauses_that_could_match<I: Interner>(
                 .to_program_clauses(builder),
         },
         DomainGoal::Holds(WhereClause::LifetimeOutlives(LifetimeOutlives { a, b })) => {
-            builder.push_clause(
-                WhereClause::LifetimeOutlives(LifetimeOutlives {
+            builder.push_fact_with_constraints(
```

Comment:
The problem with the recursive solver is that this is generating an overly specific rule. It is creating a rule that is **specific** to a, b, but we can make a more general rule like `forall<a, b> { Outlives('a: 'b) :- [] && Constraint('a: 'b)` -- sorry, not sure of the notation. =)

We do need to add a helper analogous to [`ClauseBuilder::push_bound_ty`](http://rust-lang.github.io/chalk/chalk_solve/clauses/builder/struct.ClauseBuilder.html#method.push_bound_ty) but for lifetimes. Then we can do something like this:

```rust
builder.push_bound_lifetime(|builder, a| {
  builder.push_bound_lifetime(|builder, b| {
    builder.push_fact_with_constraints(
      LifetimeOutlives { a, b },
      Some(/* new constraint that a: b */)
    )
  })
})
```

Reply:
Ah I see, that makes sense. Thanks for the help!